### DISCUSSION TITLE
Polybar support

### DISCUSSION THREAD

**Discussion by og900aero** [bakkeby/dusk#26] [Ideas]
I am very sympathetic to this wm. Any chance you will ever support Polybar?

**Comment by og900aero** [comment]
"No."

**Comment by bakkeby** [comment]
Sorry I hadn't seen this discussion.

There is no explicit or intended support for external bars such as Polybar or Lemonbar, but there are EWMH hints in place that will make it possible.

I tried polybar with the default configuration and it seems to work well. As-is it was merely placed on top of the internal bar. To run it without internal bars I think we would need something like the insets patch for dwm to reserve space for the external bar. Otherwise it looks like the programs are just cut off at the top.

This is fine in a single monitor setup, but you will likely have some trouble in a multi-monitor setup - not because of placement and configuration, but because EWMH does not have a concept of having different workspaces on a per monitor basis. There is only the concept of having different "desktops" where each desktop span all monitors. I assume that this was conjured before Xinerama and multi-monitor support was a thing.

As such you will have all workspaces listed in polybar regardless of which monitor those workspaces are on.

Overall I think using Polybar instead of the built-in bar is feasible. You will, however, lose out on a lot of bar related features like titles for windows in a deck layout for example. This may of course not be an issue for you.

**Reply by og900aero** [reply]
Whoa. Working fine my polybar config. (i dont want titles, deck layout etc. Polybar delivers a proven solution for me.)
Yes, the reserve space is the only one problem, because I dont have multi monitor setup. Only one screen (laptop).
Maybe this patch help for me the reserved space? https://dwm.suckless.org/patches/anybar/

**Reply by bakkeby** [reply]
I have some proposed changes on branch https://github.com/bakkeby/dusk/tree/externalbars if you would want to have a look.

I would need to document this and write up a guide, but the gist of it is that you can declare a bar to be external by adding the class, instance and/or window title by appending values to the `bars` array.

```c
static const BarDef bars[] = {
	/* monitor idx  vert   x     y      w     h     name              ext class   ext inst  ext name  */
	{  0,      0,   0,    "0%    0%     -1w  -1h ", "Primary top",    "Bar", "lemonbar", "bar" },
	{  0,      1,   0,    "0%    100%   -1w  -1h ", "Primary bottom", "Polybar",  NULL,     "polybar-example_DP-1" },
	{  1,      0,   0,    "0%    0%     100% -1h ", "Secondary top" },
	{  1,      1,   0,    "0%    100%   100% -1h ", "Secondary bottom" },
	{  2,      0,   0,    "0%    0%     100% -1h ", "Tertiary top" },
	{  2,      1,   0,    "0%    100%   100% -1h ", "Tertiary bottom" },
	{  3,      0,   0,    "0%    0%     100% -1h ", "Fourth top" },
	{  3,      1,   0,    "0%    100%   100% -1h ", "Fourth bottom" },
};
```

If set up correctly you should be able to toggle the display of the bar using Super+b as normal, and the window area should automatically be adjusted when you start or kill the external bar window.

Using -1w and -1h in the size and position bar config makes it so that it will take the height and width from the bar window itself (as in this is controlled via Polybar config).

As for the bar position - this will have to be controlled by the window manager using e.g. the 0% 0% settings, for now at least.

If you enable debug logging then dusk will log the class, instance and title of the window that is being checked against the bar filters. Can be useful if you are having trouble matching the external bar window.

**Reply by og900aero** [reply]
Yes, I want to try. How can I use externalbars of branch for test?

**Reply by bakkeby** [reply]
If you do a `git pull` where you checked dusk out then you can do a `git checkout externalbars`  to change to that branch.

(and you do a `git checkout master` to get back)

**Reply by og900aero** [reply]
Yoo-hoo it works perfectly! :) Show polybar, works fine functions, workspaces, gaps etc. And works hide polybar too (super+b).

**Reply by bakkeby** [reply]
I have merged these changes now and I also added a guide for reference:
https://github.com/bakkeby/dusk/wiki/How-to-set-up-external-bars

**Reply by og900aero** [reply]
Thx. I start custom config from previous WM (spectrwm, dwm, bspwm).

**Comment by og900aero** [comment]
I want to pay my respects because this wm is fantastic. It has a lot of features, it's configurable, easy to understand, and what you can't do the first time you can understand from the well put together very good documentation. Thank you very much for your work!
